compressed_depth_image_transport/src/compressed_depth_publisher.cpp

```diff
@@ -164,5 +173,54 @@ void CompressedDepthPublisher::declareParameter(
     RCLCPP_DEBUG(logger_, "%s was previously declared", definition.descriptor.name.c_str());
     param_value = node_->get_parameter(param_name).get_parameter_value();
   }
+
+  if (node_->get_effective_namespace().length() > 1) {
+    // deprecated parameters starting with the dot character (e.g. .image_raw.compressed.format)
+    const std::string deprecated_dot_name = "." + base_name + "." + transport_name + "." +
+      definition.descriptor.name;
+    deprecated_parameters_.insert(deprecated_dot_name);
+
+    try {
+      node_->declare_parameter(deprecated_dot_name, param_value, definition.descriptor);
+    } catch (const rclcpp::exceptions::ParameterAlreadyDeclaredException &) {
+      RCLCPP_DEBUG(logger_, "%s was previously declared", definition.descriptor.name.c_str());
+    }
+  }
+}
+
+void CompressedDepthPublisher::preSetParametersCallback(
+  std::vector<rclcpp::Parameter> & parameters, std::string base_name)
+{
+  for (auto & param : parameters) {
+    // Check if parameter is deprecated
+    if (deprecated_parameters_.find(param.get_name()) != deprecated_parameters_.end()) {
+      // name was generated from base_name, has to succeed
+      size_t base_name_index = param.get_name().find(base_name);
+      size_t param_name_index = base_name_index + base_name.size();
+
+      // Check if scoped parameter name
+      if (param.get_name().substr(param_name_index + 1, getTransportName().size()) ==
+        getTransportName())
+      {
+        param_name_index += getTransportName().size() + 1;
+      }
+
+      std::string recommended_name = base_name + "." + getTransportName() + "." +
+        param.get_name().substr(param_name_index + 1);
+
+      rclcpp::Parameter recommended_value = node_->get_parameter(recommended_name);
+
+      // do not emit warnings if deprecated value matches
+      if (param.get_value_message() == recommended_value.get_value_message()) {
+        continue;
+      }
+
+      RCLCPP_WARN_STREAM(logger_,
+          "parameter `" << param.get_name() << "` is deprecated; use canonical name: `" <<
+          recommended_name << "`");
+
+      parameters.push_back(rclcpp::Parameter(recommended_name, param.get_parameter_value()));
```

Comment:
Modifying the vector while iterating can cause undefined behavior. Consider creating a separate ``std::vector<rclcpp::Parameter>`` to store the new parameters during this iteration, and inserting them to the end of ``parameter`` afterwards.